src/lib/sign_munge.c

```diff
@@ -0,0 +1,244 @@
+    if (!(sm->munge = munge_ctx_create ()))
+        goto error;
+    if (flux_security_aux_set (ctx, auxname, sm,
+                               (flux_security_free_f)sm_destroy) < 0)
+        goto error;
+    sm->max_ttl = cf_int64 (cf_get_in (cf, "max-ttl"));
+    if ((munge_config = cf_get_in (cf, "munge"))) {
+        struct cf_error cfe;
+        const cf_t *entry;
+        if (cf_check (munge_config, munge_opts, CF_STRICT, &cfe) < 0) {
+            security_error (ctx, "sign-munge-init: %s", cfe.errbuf);
+            goto error_nomsg;
+        }
+        if ((entry = cf_get_in (munge_config, "socket-path")))
+            socket_path = cf_string (entry);
+    }
+    if (socket_path) {
+        munge_err_t e;
+        e = munge_ctx_set (sm->munge, MUNGE_OPT_SOCKET, socket_path);
+        if (e != EMUNGE_SUCCESS) {
+            security_error (ctx, "sign-munge-init: munge_opt_set %s: %s",
+                            socket_path, munge_ctx_strerror (sm->munge));
+            goto error_nomsg;
+        }
+    }
+    return 0;
+error:
+    security_error (ctx, NULL);
+error_nomsg:
+    sm_destroy (sm);
+    return -1;
+}
+
+/* Compute hash over HEADER.PAYLOAD (input), then "sign" the hash,
+ * producing a munge credential.
+ * Reserve first byte of munge payload to indicate which hash algorithm.
+ */
+static const char *op_sign (flux_security_t *ctx, const char *input,
+                              int flags)
+{
+    struct sign_munge *sm = flux_security_aux_get (ctx, auxname);
+    BYTE digest[SHA256_BLOCK_SIZE + 1] = { HASH_TYPE_SHA256 };
+    SHA256_CTX shx;
+    char *cred;
+    munge_err_t e;
+
+    assert (sm != NULL);
+    sha256_init (&shx);
+    sha256_update (&shx, (const BYTE *)input, strlen (input));
+    sha256_final (&shx, digest + 1);
+    e = munge_encode (&cred, sm->munge, digest, sizeof (digest));
+    if (e != EMUNGE_SUCCESS) {
+        errno = EINVAL;
+        security_error (ctx, "sign-munge-sign: %s",
+                        munge_ctx_strerror (sm->munge));
+        return NULL;
+    }
+    free (sm->cred_cache);
+    sm->cred_cache = cred;
+    return sm->cred_cache;
+}
+
+/* Recompute hash over HEADER.PAYLOAD portion of input, then munge_decode
+ * the SIGNATURE portion of input as a munge cred, and check:
+ * - munge cred's payload matches the computed hash
+ * - security header userid matches munge cred uid
+ * - munge encode time plus configured max-ttl is not past.
+ */
+static int op_verify (flux_security_t *ctx, const char *input,
+                        const struct kv *header, int flags)
+{
+    struct sign_munge *sm = flux_security_aux_get (ctx, auxname);
+    char *q;
+    munge_err_t e;
+    char *indigest = NULL;
+    int indigestsz = 0;
+    uid_t uid;
+    uint64_t userid;
+    time_t now;
+    time_t encode_time;
+    int saved_errno;
+
+    assert (sm != NULL);
+
+    /* HEADER.PAYLOAD.SIGNATURE
+     * ^input        ^q
+     */
+    q = strrchr (input, '.');
+    assert (q != NULL);
```

Comment:
No, sign.c::flux_sign_unwrap() decoded the HEADER and PAYLOAD and bails out on error prior to calling mech->verify, so input will always have at least one of the required periods.

Now that you metion it though, it seems like mech->verify should not be expected to parse HEADER.PAYLOAD.SIGNATURE.  I should change the args so that instead of 
```c
/* verify (required)
 * Given HEADER.PAYLOAD.SIGNATURE and parsed 'header', validate that
 * SIGNATURE is valid over HEADER.PAYLOAD.
 * Return 0 on success, or -1 on error with errno and context error set.
 */
typedef int (*sign_mech_verify_f)(flux_security_t *ctx, const char *input,
                                  const struct kv *header, int flags);
```
we have something like:
```c
/* Validate that 'signature' (a NULL terminated string) is valid over
 * 'input' of size 'inputsz'.
 */
typedef int (*sign_mech_verify_f)(flux_security_t *ctx,
                                  const char *input, int inputsz,
                                  const char *signature,
                                  const struct kv *header, int flags);
```

Reply:
Similar check is in `header_decode`:
```C
    if (!(p = strchr (input, '.'))) {
        errno = EINVAL;
        return NULL;
    }
    src = input;
    srclen = p - input;

```

Does this need to handle corner cases (period at beginning or end of `input`)?

Reply:
I think these are OK and are covered by lib/test/sign.c, e.g.

Period at the end "HEADER.PAYLOAD."
> flux_unwrap_fails on missing SIGNATURE with EINVAL

Period at the beginning ".PAYLOAD.SIGNATURE"
> flux_unwrap_fails on empty HEADER with EINVAL